### PR TITLE
feat: add sorting feature and related conf

### PR DESCRIPTION
**Pull request by Helias**
closes https://github.com/azerothcore/mod-transmog/issues/133

I tested it in-game with few items

### CODE REVIEW

**Comment by Nyeriah** [comment]
@Ceikry

File: src/transmog_scripts.cpp [hunk 1]
```diff
@@ -1201,7 +1201,17 @@ class WS_Transmogrification : public WorldScript
...
             LOG_INFO("module", "Loading transmog appearance collection cache....");
             uint32 collectedAppearanceCount = 0;
-            QueryResult result = CharacterDatabase.Query("SELECT account_id, item_template_id FROM custom_unlocked_appearances");
+
+
+            std::string query = "SELECT account_id, item_template_id FROM custom_unlocked_appearances";
+
+            if (sConfigMgr->GetOption<bool>("Transmogrification.EnableSortByQualityAndName", true)) {
+                const std::string acore_world_name = sConfigMgr->GetOption<std::string>("Transmogrification.AcoreWorldName", "acore_world");
+                query = "SELECT custom_unlocked_appearances.account_id, custom_unlocked_appearances.item_template_id, " + acore_world_name + ".item_template.name, " + acore_world_name + ".item_template.Quality FROM custom_unlocked_appearances INNER JOIN " + acore_world_name + ".item_template ON custom_unlocked_appearances.item_template_id=" + acore_world_name + ".item_template.entry ORDER BY " + acore_world_name + ".item_template.Quality DESC, " + acore_world_name + ".item_template.name ASC;";
```
Comment (Nyeriah) [inline src/transmog_scripts.cpp:1210]: This query will perform very badly with this cross-db join

Comment (Helias) [reply src/transmog_scripts.cpp:1210]: I know, I do not know how to improve it
would be better to convert it into a left join?

**Comment by Ceikry** [comment]
Couldn't you just sort the generated vector of allowed items instead? Sure you sort multiple times but it's a cheap sort on such a small collection and wouldn't be nearly as messy.

something like this (you can and should probably make the name comparison better)
```cpp
bool CmpTmog (Item* i1, Item* i2)
{
    const ItemTemplate* i1t = i1->GetTemplate();
    const ItemTemplate* i2t = i2->GetTemplate();
    return i1t->Name1[0] > i2t->Name1[0] || i1t->Quality > i2t->Quality;
}

    sort(allowedItems.begin(), allowedItems.end(), CmpTmog);
```

**Comment by Ceikry** [comment]
> Couldn't you just sort the generated vector of allowed items instead? Sure you sort multiple times but it's a cheap sort on such a small collection and wouldn't be nearly as messy.
> 
> something like this (you can and should probably make the name comparison better)
> 
> ```c++
> bool CmpTmog (Item* i1, Item* i2)
> {
>     const ItemTemplate* i1t = i1->GetTemplate();
>     const ItemTemplate* i2t = i2->GetTemplate();
>     return i1t->Name1[0] > i2t->Name1[0] || i1t->Quality > i2t->Quality;
> }
> 
>     sort(allowedItems.begin(), allowedItems.end(), CmpTmog);
> ```

in fact I would advise mildly reworking this to use string::compare for the name comparisons.

**Comment by Helias** [comment]
I will try that, thank you!

**Comment by Helias** [comment]
I restored the code as it was before and added that `sort(allowedItems.begin(), allowedItems.end(), CmpTmog);` before `return allowedItems` but it seems not working

**Comment by Helias** [comment]
if I remove `i1t->Quality > i2t->Quality` it orders the items per name, so I think that in some way I could sort them, I will look better into this, thanks a lot for the input

**Comment by Ceikry** [comment]
Yeah you might have to tweak the conditions for the comparison to be a bit more intelligent, I just kinda slapped an example together from what I remembered about C++ sorting

**Comment by Helias** [comment]
now it works
image
image
image
image

File: src/transmog_scripts.cpp [hunk 2]
```diff
@@ -1201,7 +1215,9 @@ class WS_Transmogrification : public WorldScript
         {
             LOG_INFO("module", "Loading transmog appearance collection cache....");
             uint32 collectedAppearanceCount = 0;
+
             QueryResult result = CharacterDatabase.Query("SELECT account_id, item_template_id FROM custom_unlocked_appearances");
+
             if (result)
```
Comment (Helias) [inline src/transmog_scripts.cpp:1221]:
```suggestion
            uint32 collectedAppearanceCount = 0;
            QueryResult result = CharacterDatabase.Query("SELECT account_id, item_template_id FROM custom_unlocked_appearances");
            if (result)
```

**Review by Nyeriah — APPROVED** [review]

**Comment by Nyeriah** [comment]
Way much better, good job! 👍

**Comment by Helias** [comment]
thanks you both!

**Pull request merged**